### ISSUE TITLE
useReducer causing loop when used with context, but useState is not

### ISSUE THREAD

**Issue by bradwestfall** [facebook/react#16537]
**Do you want to request a *feature* or report a *bug*?**

Bug, I think

**What is the current behavior?**

I apologize this might seem convoluted, but there are several things involved that make this problem:

1. I have a typical Context provider that stored it's state values using `useReducer`
2. In a leaf component where I'm consuming that context, I have an effect that does some async code and then sets state in the context (setState function passed down through context)

```js
function ShoppingPage() {
  const { shoppingCartState } = useContext(Context);
  useEffect(() => {
    getShoppingCartStuff().then(shoppingCartState.setCart);
  }, []);
  // I later added [shoppingCartState.setCart] to the dep array. See comments below

  return <div>Cart: {shoppingCartState.cart}</div>;
}
```

3. Recently I added the "exhaustive-deps" lint rules and caught that I wasn't adding `shoppingCartState.setCart` to that dep array. So I added it and started getting an infinite loop. My thought process was `shoppingCartState` must be getting a new identity in memory. 
4. After exploring for a while, I decided to change the state of the context to use `useState` and the infinite loop went away.

So just to be clear, the state in context being managed with `useState` or `useReducer` doesn't cause an infinite loop until I add that dep array value for `shoppingCartState.setCart`, then `useReducer` causes a loop and `useState` doesn't

Example with both `useState` and `useReducer` versions
https://codesandbox.io/s/competent-boyd-yzzyh

**What is the expected behavior?**

Hopefully no loop

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

16.8 and 16.9 tested

**Comment by dmytro-lymarenko** [comment]
The problem here
```js
function useShoppingCartState() {
  const [state, dispatch] = useReducer((state, action) => {
    switch (action.type) {
      case "SET_CART": {
        return { ...state, cart: action.cart };
      }
      default:
        return state;
    }
  }, initialState);

  function setCart(cart) {
    // console log, count, and condition are just here to show
    // there's a loop and to stop it
    console.log("this should not happen more than once", count);
    count++;
    if (count < 10) {
      dispatch({ type: "SET_CART", cart });
    }
  }

  return {
    ...state,
    setCart
  };
}
```
is that you don't memoize `setCart` function with useCallback that is why that effect gets new function every render which causes it to fall into infinite loop.
Another way to avoid infinite loop is to use `dispatch` function directly in `useEffect`:
```js
useEffect(() => {
    getShoppingCartStuff().then(cart => {
        shoppingCartState.dispatch({ type: "SET_CART", cart });
    });
  }, [shoppingCartState.dispatch]);
```
and to finish it I suggest to add `canceled` flag:
```js
useEffect(() => {
    let canceled = false;
    getShoppingCartStuff().then(cart => {
        if (!canceled) {
            shoppingCartState.dispatch({ type: "SET_CART", cart });
        }
    });
    return () => {
        canceled = true;
    };
  }, [shoppingCartState.dispatch]);
```

**Comment by dmytro-lymarenko** [comment]
This
> After exploring for a while, I decided to change the state of the context to use useState and the infinite loop went away.

works because setState is already memoized ;)

**Comment by bradwestfall** [comment]
Cool, I tried memoizing (useCallback-ing) the wrong thing when I was trying to figure it out. This makes sense now. Also, yep, I left out the cleanup to be slightly more to-the-point

**Comment by Satsh24** [comment]
Hey @bradwestfall @dmytro-lymarenko I have a similar issue and I'm using useState not sure how to get around this. Please suggest.
Here is the code Sanbox
[https://codesandbox.io/s/red-shadow-jsqfx]
Thanks

**Comment by bradwestfall** [comment]
@Satsh24 Seems like it might be a case of `fetchBooks` inside your context provider not being memoized. You're custom provider is a component just like any other, that experiences re-renders when it's state changes - which means the `const fetchBooks = () => {}` function is being re-created (a new identity in memory) every time. Then when the linter rule asks for that function to be in it's dependency array, then on a successful resolution it sets state in the provider, causing a new version of fetchBooks to be created which means the thing you pass into the dependency array is changed which means it fetches again (loops). Wrap the fetchBooks in a useCallback. Keep in mind useMemo is for memoizing the "values" of functions in a component and useCallback is for memoizing the functions themselves. I wrote something about all this you might like https://reacttraining.com/blog/when-to-use-functions-in-hooks-dependency-array/

And btw, when I started this issue I _thought_ I was memoizing the right thing but I wasn't so this is a bad issue
